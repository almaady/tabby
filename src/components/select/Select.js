/* eslint-disable react/sort-comp */
import React from "react";
import PropTypes from "prop-types";
import classNames from "classnames";
import SelectHeader from "./SelectHeader";
import SelectList from "./SelectList";

/**
 * @class Select - It is the wrapper for the Select Header and the Select List
 */
class Select extends React.PureComponent {
  /**
   * @property {array} options - List of options
   * @property {string} placeholder - Placeholder for the select
   * @property {string} icon - Name of the icon if the placeholder needs one
   * @property {string} className - ClassName if you need to customize
   * @property {func} onSelect - Function triggered by the item
   */
  constructor(props) {
    super(props);
    const { selectedValue } = this.props;
    this.selectRef = React.createRef();
    this.state = {
      listOpen: false,
      selectedValue
    };
  }

  static propTypes = {
    // eslint-disable-next-line react/forbid-prop-types
    options: PropTypes.array.isRequired,
    placeholder: PropTypes.string.isRequired,
    icon: PropTypes.string,
    className: PropTypes.string,
    onSelect: PropTypes.func.isRequired,
    selectedValue: PropTypes.string
  };

  componentDidMount() {
    document.addEventListener("mousedown", this.handleClickOutside);
  }

  componentWillUnmount() {
    document.removeEventListener("mousedown", this.handleClickOutside);
  }

  toggleList = () => {
    this.setState(prevState => ({
      listOpen: !prevState.listOpen
    }));
  };

  handleClickOutside = event => {
    const { listOpen } = this.state;
    if (
      this.selectRef.current &&
      !this.selectRef.current.contains(event.target) &&
      listOpen
    ) {
      this.setState({
        listOpen: false
      });
    }
  };

  getSelectedItem = value => {
    const { options } = this.props;
    const item = options.find(_item => _item.value === value);
    return item;
  };

  _setItemSelected = itemValue => {
    this.setState({
      selectedValue: itemValue
    });
  };

  selectItemByValue = itemValue => {
    const item = this.getSelectedItem(itemValue);
    this._setItemSelected(itemValue);
    this.props.onSelect(item);
    this.toggleList();
  };

  selectItem = event => {
    const itemValue = event.target.dataset.value;
    this.selectItemByValue(itemValue);
  };

  componentDidUpdate(prevProps) {
    const { selectedValue } = this.props;
    if (prevProps.selectedValue !== selectedValue) {
      this._setItemSelected(selectedValue);
    }
  }

  render() {
    const { className, options, icon, placeholder } = this.props;
    const { listOpen, selectedValue } = this.state;
    const selectedItem = this.getSelectedItem(selectedValue);

    return (
      <div
        className={classNames("select__wrapper", className)}
        ref={this.selectRef}
      >
        <SelectHeader
          title={selectedItem ? selectedItem.title : placeholder}
          icon={selectedItem ? selectedItem.icon : icon}
          onClick={this.toggleList}
          listOpen={listOpen}
        />
        {listOpen && (
          <SelectList
            options={options}
            onClick={this.selectItem}
            selectedValue={selectedValue}
          />
        )}
      </div>
    );
  }
}

Select.defaultProps = {
  className: "",
  icon: "",
  selectedValue: ""
};

export default Select;
