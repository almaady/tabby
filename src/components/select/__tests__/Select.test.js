import React from "react";
import { shallow, mount } from "enzyme";
import sinon from "sinon";

import Select from "../Select";

describe("Select", () => {
  let props;
  beforeEach(() => {
    props = {
      placeholder: "Select",
      options: [
        {
          title: "Option 1",
          icon: "top",
          value: "option 1"
        },
        {
          title: "Option 2",
          icon: "left",
          value: "option 2"
        },
        {
          title: "Option 3",
          icon: "right",
          value: "option 3"
        }
      ],
      onSelect: () => {}
    };
  });

  it("should render correctly with default state", () => {
    const component = shallow(<Select {...props} />);
    expect(component).toMatchSnapshot();
  });
  it("should change state when you click on SelectHeader", () => {
    const click = sinon.spy();
    const component = shallow(<Select {...props} onClick={click} />);
    component.find("SelectHeader").simulate("click");
    expect(component.state().listOpen).toEqual(true);
  });
  it("should use selected item as param", () => {
    const click = sinon.spy();
    const component = mount(<Select {...props} onSelect={click} />);
    component.find("SelectHeader").simulate("click");
    component.find({ value: "option 2" }).simulate("click");
    expect(click.calledWithExactly(props.options[1])).toBeTruthy();
  });
  it("should select one option from the list", () => {
    const click = sinon.spy();
    const component = mount(<Select {...props} onClick={click} />);
    component.find("SelectHeader").simulate("click");
    component.find({ value: "option 2" }).simulate("click");
    expect(component.state().selectedValue).toEqual("option 2");
  });
  it("should change title when you selected one option", () => {
    const component = mount(<Select {...props}  />);
    component.find("SelectHeader").simulate("click");
    component.find({ value: "option 2" }).simulate("click");
    expect(component.find("SelectHeader").props().title).toEqual("Option 2")
  });
  it("should re render when props change", () => {
    let component = mount(<Select {...props} />);
    component.setProps({selectedValue: props.options[1].value })
    expect(component.state().selectedValue).toEqual(props.options[1].value);
  });
  it("should change state when you click outside", () => {
    const wrapper = mount(
      <div>
        <Select {...props} />
      </div>
    );
    wrapper.find("SelectHeader").simulate("click");
    const event = new MouseEvent("mousedown");
    document.dispatchEvent(event);
    expect(wrapper.find(Select).state().listOpen).toEqual(false);
  });
  
});
