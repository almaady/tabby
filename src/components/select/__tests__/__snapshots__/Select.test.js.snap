// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select should render correctly with default state 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Select
    className=""
    icon=""
    onSelect={[Function]}
    options={
      Array [
        Object {
          "icon": "top",
          "title": "Option 1",
          "value": "option 1",
        },
        Object {
          "icon": "left",
          "title": "Option 2",
          "value": "option 2",
        },
        Object {
          "icon": "right",
          "title": "Option 3",
          "value": "option 3",
        },
      ]
    }
    placeholder="Select"
    selectedValue=""
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <SelectHeader
          className=""
          icon=""
          listOpen={false}
          onClick={[Function]}
          title="Select"
        />,
        false,
      ],
      "className": "select__wrapper",
    },
    "ref": Object {
      "current": null,
    },
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "className": "",
          "icon": "",
          "listOpen": false,
          "onClick": [Function],
          "title": "Select",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      false,
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <SelectHeader
            className=""
            icon=""
            listOpen={false}
            onClick={[Function]}
            title="Select"
          />,
          false,
        ],
        "className": "select__wrapper",
      },
      "ref": Object {
        "current": null,
      },
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "className": "",
            "icon": "",
            "listOpen": false,
            "onClick": [Function],
            "title": "Select",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        false,
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    Symbol(enzyme.__providerValues__): undefined,
  },
  Symbol(enzyme.__providerValues__): Map {},
  Symbol(enzyme.__childContext__): null,
}
`;
